### PR TITLE
G-API: Fix various issues for 4.2 release

### LINKED ISSUE #15836: G-API: valgrind issues (2019-11-02)
[Valgrind build](http://pullrequest.opencv.org/buildbot/builders/master_valgrind-lin64-debug/builds/10203)

```
==14190== Use of uninitialised value of size 8
==14190==    at 0x6F1A811: _itoa_word (_itoa.c:180)
==14190==    by 0x6F1C1F7: printf_positional (vfprintf.c:2022)
==14190==    by 0x6F1ECB9: vfprintf (vfprintf.c:1688)
==14190==    by 0x6FF4168: __vsnprintf_chk (vsnprintf_chk.c:63)
==14190==    by 0x6FF4094: __snprintf_chk (snprintf_chk.c:34)
==14190==    by 0x5BAF4E: snprintf (stdio2.h:65)
==14190==    by 0x5BAF4E: testing::(anonymous namespace)::PrintByteSegmentInObjectTo(unsigned char const*, unsigned long, unsigned long, std::ostream*) (ts_gtest.cpp:10840)
==14190==    by 0x5BDA53: PrintBytesInObjectToImpl (ts_gtest.cpp:10858)
==14190==    by 0x5BDA53: testing::internal2::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (ts_gtest.cpp:10880)
==14190==    by 0x59491C: PrintValue (ts_gtest.h:10569)
==14190==    by 0x59491C: operator<< <char, std::char_traits<char>, cv::gapi::GKernelPackage> (ts_gtest.h:10660)
==14190==    by 0x59491C: DefaultPrintNonContainerTo<cv::gapi::GKernelPackage> (ts_gtest.h:10701)
==14190==    by 0x59491C: DefaultPrintTo<cv::gapi::GKernelPackage> (ts_gtest.h:10892)
==14190==    by 0x59491C: PrintTo<cv::gapi::GKernelPackage> (ts_gtest.h:10927)
==14190==    by 0x59491C: Print (ts_gtest.h:11182)
==14190==    by 0x59491C: UniversalPrint<cv::gapi::GKernelPackage> (ts_gtest.h:11373)
==14190==    by 0x59491C: Print (ts_gtest.h:11301)
==14190==    by 0x59491C: PrintToString<cv::gapi::GKernelPackage> (ts_gtest.h:11510)
==14190==    by 0x59491C: testing::internal::ParameterizedTestCaseInfo<opencv_test::(anonymous namespace)::GAPI_Streaming>::RegisterTests() (ts_gtest.h:12100)
==14190==    by 0x5BC36E: RegisterTests (ts_gtest.h:12228)
==14190==    by 0x5BC36E: testing::internal::UnitTestImpl::RegisterParameterizedTests() (ts_gtest.cpp:4132)
==14190==    by 0x5D6744: testing::internal::UnitTestImpl::PostFlagParsingInit() (ts_gtest.cpp:6463)
==14190==    by 0x5D683B: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (ts_gtest.cpp:7430)
==14190==    by 0x5D6884: testing::InitGoogleTest(int*, char**) (ts_gtest.cpp:7448)
==14190==  Uninitialised value was created by a stack allocation
==14190==    at 0x580D6F: opencv_test::gtest_TestStreamingGAPI_Streaming_EvalGenerator_() (gapi_streaming_tests.cpp:305)
```

### LINKED ISSUE #15882: G-API: coverage failure
[Build](http://pullrequest.opencv.org/buildbot/builders/master_coverage-lin64-debug/builds/10980):

Test: `TestStreaming/GAPI_Streaming.SmokeTest_VideoConstSource_NoHang/1`

```
[ RUN      ] TestStreaming/GAPI_Streaming.SmokeTest_VideoConstSource_NoHang/1, where GetParam() = 80-byte object <50-CC 2E-DF D4-55 00-00 A7-00 00-00 00-00 00-00 30-E9 31-DF D4-55 00-00 5E-00 00-00 00-00 00-00 00-00 80-3F D4-55 00-00 A7-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/passes/kernels.cpp (187) expandKernels Final graph: 3 nodes

[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/gcompiler.cpp (372) runPasses All compiler passes are successful
[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/backends/fluid/gfluidbackend.cpp (851) GFluidExecutable Initializing 2 fluid buffer(s)

[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/backends/fluid/gfluidbackend.cpp (918) GFluidExecutable Internal buffers: 0.00 KB

[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/passes/kernels.cpp (187) expandKernels Final graph: 8 nodes

[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/passes/exec.cpp (581) fuseGeneral island_#0x55d4dfb4ee00 can be merged into island_#0x55d4dfc103e0
[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/passes/exec.cpp (601) fuseGeneral merge(island_#0x55d4dfc103e0,island_#0x55d4dfb4ee00) was successful!
[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/compiler/gcompiler.cpp (372) runPasses All compiler passes are successful
[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/backends/fluid/gfluidbackend.cpp (851) GFluidExecutable Initializing 3 fluid buffer(s)

[ INFO:0] global /build/master_coverage-lin64-debug/opencv/modules/gapi/src/backends/fluid/gfluidbackend.cpp (918) GFluidExecutable Internal buffers: 0.00 KB

/build/master_coverage-lin64-debug/opencv/modules/gapi/test/streaming/gapi_streaming_tests.cpp:302: Failure
Expected equality of these values:
  ref_frames
    Which is: 100
  test_frames
    Which is: 99
[  FAILED  ] TestStreaming/GAPI_Streaming.SmokeTest_VideoConstSource_NoHang/1, where GetParam() = 80-byte object <50-CC 2E-DF D4-55 00-00 A7-00 00-00 00-00 00-00 30-E9 31-DF D4-55 00-00 5E-00 00-00 00-00 00-00 00-00 80-3F D4-55 00-00 A7-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00> (2591 ms)
```

### LINKED ISSUE #15903: G-API: build warnings with GCC 9
```
build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp: In function 'std::unique_ptr<cv::gapi::fluid::BufferStorage> cv::gapi::fluid::{anonymous}::createStorage(int, int, int, int, cv::gapi::fluid::BorderOpt)':
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:350:25: warning: redundant move in return statement [-Wredundant-move]
  350 |         return std::move(storage);
      |                ~~~~~~~~~^~~~~~~~~
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:350:25: note: remove 'std::move' call
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:355:21: warning: redundant move in return statement [-Wredundant-move]
  355 |     return std::move(storage);
      |            ~~~~~~~~~^~~~~~~~~
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:355:21: note: remove 'std::move' call
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp: In function 'std::unique_ptr<cv::gapi::fluid::BufferStorage> cv::gapi::fluid::{anonymous}::createStorage(const cv::gapi::own::Mat&, cv::gapi::own::Rect)':
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:363:21: warning: redundant move in return statement [-Wredundant-move]
  363 |     return std::move(storage);
      |            ~~~~~~~~~^~~~~~~~~
/build/precommit_custom_linux/opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:363:21: note: remove 'std::move' call
```

can be validated through Fedora 31 build image.

```
force_builder=Custom
buildworker:Custom=linux-1
build_image:Custom=fedora:31
```

### PR DESCRIPTION
**Pull request by dmatveev**
- Coverity issues (`GModel` inefficiencies);
- Fixes #15903
- Likely fixes #15882 ([validated](http://pullrequest.opencv.org/buildbot/builders/master_coverage-lin64-debug/builds/11022) :+1: ) 
- Likely fixes #15836 ([validated](http://pullrequest.opencv.org/buildbot/builders/master_valgrind-lin64-debug/builds/10209) :+1: )

```
buildworker:Custom=linux-1
build_image:Custom=fedora:31
```

### CODE REVIEW

**Comment by dmatveev** [comment]
@AsyaPronina @rgarnov  please review the latest fixes.

@alalek can you please trigger the relevant builds to test the fixed issues?

File: modules/gapi/src/executor/gstreamingexecutor.cpp [hunk 1]
```diff
@@ -132,6 +132,177 @@ void sync_data(cv::GRunArgs &results, cv::GRunArgsP &outputs)
...
+// graph island in its own thread, and threads communicate via bounded
+// (limited in size) queues.  Threads poll their input queues in the
+// infinite loops and pass the data to their Island executables when
+// the full input vector (a "stack frame") arrives.
+//
+// If the input stream is over or stop() is called, "Stop" messages
+// are broadcasted in the graph from island to island via queues,
+// starting with the emitters (sources). Since queues are bounded,
+// thread may block on push() if the queue is full already and is not
+// popped for some reason in the reader thread. In order to aovid
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/src/executor/gstreamingexecutor.cpp]: Thanks

File: modules/gapi/src/executor/gstreamingexecutor.cpp [hunk 2]
```diff
@@ -132,6 +132,177 @@ void sync_data(cv::GRunArgs &results, cv::GRunArgsP &outputs)
...
+//
+// "Stop" messages are sent to the pipeline in these three cases:
+// 1. User has called stop(): a "Stop" message is sent to every input
+//    queue.
+// 2. Input video stream has reached its end -- its emitter sends Stop
+//    to its readers AND asks constant emitters (emitters attached to
+//    const data -- infinite data generators) to push Stop messages as
+//    well - in order to maintain a regular Stop procedure as defined
+//    above.
+// 3. "Sop" message coming from a constant emitter after triggering an
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/src/executor/gstreamingexecutor.cpp]: ...

File: modules/gapi/src/executor/gstreamingexecutor.cpp [hunk 3]
```diff
@@ -132,6 +132,177 @@ void sync_data(cv::GRunArgs &results, cv::GRunArgsP &outputs)
...
+//    EOS notification -- see (2).
+//
+// There is a problem with (3). Sometimes it terminates the pipeline
+// too early while some frames could still be produced with no issue,
+// and our test fail with error like "got 99 frames, expected 100".
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/src/executor/gstreamingexecutor.cpp]: :)

File: modules/gapi/src/executor/gstreamingexecutor.cpp [hunk 4]
```diff
@@ -132,6 +132,177 @@ void sync_data(cv::GRunArgs &results, cv::GRunArgsP &outputs)
...
+// This is how it reproduces:
+//
+//                   q1
+//   [const input]   -----------------------> [ ISL2 ] --> [output]
+//                   q0             q2    .->
+//   [stream input]  ---> [ ISL1 ] -------'
+//
+// Video emitter is pushing frames to q0, and ISL1 is takin every
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/src/executor/gstreamingexecutor.cpp]: ._.

**Comment by dmatveev** [comment]
@rgarnov all fixed, thanks!

**Comment by dmatveev** [comment]
@alalek can this be merged? Anything else required?

**Comment by dmatveev** [comment]
Thanks!

**Comment by mshabunin** [comment]
Looks like this patch causes build errors with GCC 4.8.5 (CentOS 7):
```
../opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp: In function 'std::unique_ptr<cv::gapi::fluid::BufferStorage> cv::gapi::fluid::{anonymous}::createStorage(int, int, int, int, cv::gapi::fluid::BorderOpt)':
../opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:350:16: error: cannot bind 'std::unique_ptr<cv::gapi::fluid::BufferStorageWithBorder>' lvalue to 'std::unique_ptr<cv::gapi::fluid::BufferStorageWithBorder>&&'
         return storage;
                ^
In file included from /usr/include/c++/4.8.2/memory:81:0,
                 from ../opencv/modules/core/include/opencv2/core/cvstd_wrapper.hpp:11,
                 from ../opencv/modules/core/include/opencv2/core/cvstd.hpp:81,
                 from ../opencv/modules/core/include/opencv2/core/base.hpp:58,
                 from ../opencv/modules/core/include/opencv2/core.hpp:54,
                 from ../opencv/modules/gapi/src/precomp.hpp:12,
                 from ../opencv/modules/gapi/src/backends/fluid/gfluidbuffer.cpp:8:
/usr/include/c++/4.8.2/bits/unique_ptr.h:169:2: error:   initializing argument 1 of 'std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Up, _Ep>&&) [with _Up = cv::gapi::fluid::BufferStorageWithBorder; _Ep = std::default_delete<cv::gapi::fluid::BufferStorageWithBorder>; <template-parameter-2-3> = void; _Tp = cv::gapi::fluid::BufferStorage; _Dp = std::default_delete<cv::gapi::fluid::BufferStorage>]'
  unique_ptr(unique_ptr<_Up, _Ep>&& __u) noexcept
  ^
```

**Pull request merged**